File: 第11章 竞赛趣题/11.14 拦截导弹/拦截导弹.py
'''
程序：拦截导弹
作者：苏秦@小海豚科学馆公众号
来源：图书《趣味编程三剑客：从Scratch迈向Python和C++》
'''
# 导弹的高度
daodan = [389, 207, 155, 300, 299, 170, 158, 65]
# 拦截系统
lanjie = []
    
def lanjie_daodan(h):
    '''拦截导弹并创建系统'''
    j = 0
    while j < len(lanjie):
        if h < lanjie[j] + 1:
            lanjie[j] = h
            return
        j += 1
    lanjie.append(h)

def main():
    '''拦截导弹'''
    i = 0
    while i < 8:
        h = daodan[i]
        lanjie_daodan(h)
        i += 1
    print(len(lanjie))
    
if __name__ == '__main__':
    main()

File: 第11章 竞赛趣题/11.14 拦截导弹/拦截导弹.cpp
/*
程序：拦截导弹 
作者：苏秦@小海豚科学馆公众号
来源：图书《趣味编程三剑客：从Scratch迈向Python和C++》
*/
#include <bits/stdc++.h>
using namespace std;

// 导弹的高度 
int daodan[8] = {389, 207, 155, 300, 299, 170, 158, 65};
// 拦截系统 
vector<int> lanjie;

// 拦截导弹并创建系统
void lanjie_daodan(int h) {
    int j = 0;
    while (j < lanjie.size()) {
        if (h < lanjie[j] + 1) {
            lanjie[j] = h;
            return;
    	}
        j += 1;
    }
    lanjie.push_back(h);
}

// 拦截导弹 
int main()
{
	int i = 0;
	while (i < 8) {
		int h = daodan[i];
		lanjie_daodan(h);
		i += 1;
	}
	cout << lanjie.size() << endl;
	return 0;
}
